lib/ConsumerImpl.cc

```diff
@@ -527,6 +544,10 @@ void ConsumerImpl::messageReceived(const ClientConnectionPtr& cnx, const proto::
                                 << startMessageId.value());
             return;
         }
+        if (redeliveryCount >= deadLetterPolicy_.getMaxRedeliverCount()) {
+            possibleSendToDeadLetterTopicMessages_.emplace(m.getMessageId(), std::vector<Message>{m});
+            increaseAvailablePermits(cnx);
```

Comment:
```suggestion
            if (redeliveryCount > deadLetterPolicy_.getMaxRedeliverCount()) {
                increaseAvailablePermits(cnx);
            }
```

See the Java implementation [here](https://github.com/apache/pulsar/pull/17287/files#diff-f6e4c1c4091aa10525f331e48e66b29f22b9f7987755c1b4fb887e24f198bed6).

And should we call `redeliverUnacknowledgedMessages` here as well?

Reply:
If the redeliver logic needs extra tests. We can add a TODO here and add the tests in another PR.

Reply:
> And should we call redeliverUnacknowledgedMessages here as well?

Yes, Need to call `redeliverUnacknowledgedMessages(const std::set<MessageId>& messageIds)`, I understand it wrong before.

The design of the DLQ goes like this:

When receiving the message `redeliveryCount == deadletterPolicy_.getMaxRedeliverCount`, We add it to `possibleSendToDeadLetterTopicMessages_` waiting for the user to call `nack method` or `ackTimeOut` trigger send to DLQ.

https://github.com/apache/pulsar-client-cpp/blob/e818ead391dc52ef08bbd59c748a00492dcf0cb5/lib/NegativeAcksTracker.cc#L82-L84

https://github.com/apache/pulsar-client-cpp/blob/1721e0005975bcc9cbd49566d6047760e6621a3b/lib/UnAckedMessageTrackerEnabled.cc#L68-L73

This can reduce once redeliver message from the server to the client.

But there are also unexpected scenarios, case `transaction.abort` : apache/pulsar/pull/17287, It does not go through the `redeliverUnacknowledgedMessages(const std::set<MessageId>& messageIds)` method, so need to directly judge again whether it is greater than: `if (redeliveryCount > deadLetterPolicy_.getMaxRedeliverCount())`

I originally thought the CPP client had no transactions, so it didn't need to handle the scenario. 

But I found that the user can call this method:

https://github.com/apache/pulsar-client-cpp/blob/2980bbe9d3ef47eef4a1d468c3898b5a989b4df0/include/pulsar/Consumer.h#L359

It also triggers the server to deliver messages directly without going through the logic of sending them to DLQ. 

I added back this judgment condition and added the test to cover it: `DeadLetterQueueTest.testSendDLQTriggerByRedeliverUnacknowledgedMessages`